File: tools/JavaScript/JSDemo/decoder/util/encoding.js
/**
 * Port of `com.t4login.util.EncodingUtil` (a.k.a. `t4login.util.encoding`).
 *
 * Variable-length 7-bit encoding/decoding for 32-bit ints, 64-bit longs
 * (as BigInt), and the 96-bit unscaled-decimal format.
 *
 * Sign semantics mirror the Java/C# implementation byte-for-byte:
 *   - positive ints encode in 1..5 bytes
 *   - negative ints always encode in 5 bytes
 *   - positive longs encode in 1..9 bytes
 *   - negative longs always encode in 10 bytes
 */

import { Decimal } from '../decimal.js';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const INT32_MAX = 0x7fff_ffff;
const UINT32_MASK = 0xffff_ffff; // safe as Number (< 2^53)

const UINT64_MASK = 0xffff_ffff_ffff_ffffn;
const INT64_MAX = 0x7fff_ffff_ffff_ffffn;
const BIG_2_64 = 1n << 64n;
const BIG_2_63 = 1n << 63n;

// Sentinel: -2^31 has no positive counterpart in two's-complement 32-bit,
// so the decimal codec encodes it as a 2-bit "0x03" header marker
// instead of writing its magnitude.
const INT32_MIN_VALUE = -0x8000_0000;

// ---------------------------------------------------------------------------
// 7-bit int (signed 32-bit)
// ---------------------------------------------------------------------------

/**
 * Encode a signed 32-bit integer.
 * @param {number} value
 * @returns {Uint8Array}
 */
export function encode7BitInt(value) {
  // Mask to 32-bit unsigned.
  let v = (value >>> 0); // Uint32

  if (v <= INT32_MAX) {
    const buf = [];
    while (v >= 0x80) {
      buf.push((v & 0xff) | 0x80);
      v = v >>> 7;
    }
    buf.push(v & 0xff);
    return Uint8Array.from(buf);
  } else {
    // Negative path: 5 fixed bytes with arithmetic shift behaviour.
    // Convert unsigned to signed 32 to simulate Java's signed `>>`.
    let s = v - 0x1_0000_0000; // signed 32
    const buf = new Uint8Array(5);
    buf[0] = (s & 0xff) | 0x80;
    s = s >> 7; // JS `>>` is arithmetic on 32-bit signed
    buf[1] = (s & 0xff) | 0x80;
    s = s >> 7;
    buf[2] = (s & 0xff) | 0x80;
    s = s >> 7;
    buf[3] = (s & 0xff) | 0x80;
    s = s >> 7;
    buf[4] = s & 0x0f;
    return buf;
  }
}

/**
 * Decode a 7-bit-encoded signed 32-bit integer from a stream.
 * @param {{readByte(): number}} stream  CountingInputStream or ByteReader
 * @returns {number}
 */
export function decode7BitInt(stream) {
  let count = 0;
  let shift = 0;
  while (true) {
    const b = stream.readByte();
    count |= (b & 0x7f) << shift;
    shift += 7;
    if ((b & 0x80) === 0) break;
  }
  // Sign-extend from 32-bit. JS bitwise ops already operate on 32-bit
  // signed, so a final `| 0` is sufficient.
  return count | 0;
}

// ---------------------------------------------------------------------------
// 7-bit long (signed 64-bit, BigInt)
// ---------------------------------------------------------------------------

/**
 * Encode a signed 64-bit integer (BigInt).
 * @param {bigint} value
 * @returns {Uint8Array}
 */
export function encode7BitLong(value) {
  let v = BigInt.asUintN(64, value);

  if (v <= INT64_MAX) {
    const buf = [];
    while (v >= 0x80n) {
      buf.push(Number(v & 0xffn) | 0x80);
      v = v >> 7n;
    }
    buf.push(Number(v & 0xffn));
    return Uint8Array.from(buf);
  } else {
    // Negative: 10 fixed bytes.
    let s = v - BIG_2_64; // signed 64
    const buf = new Uint8Array(10);
    for (let i = 0; i < 9; i++) {
      buf[i] = Number(s & 0xffn) | 0x80;
      s = s >> 7n; // BigInt `>>` is arithmetic
    }
    buf[9] = Number(s & 0x0fn);
    return buf;
  }
}

/**
 * Decode a 7-bit-encoded signed 64-bit integer to a BigInt.
 * @param {{readByte(): number}} stream
 * @returns {bigint}
 */
export function decode7BitLong(stream) {
  let count = 0n;
  let shift = 0n;
  while (true) {
    const b = stream.readByte();
    count |= BigInt(b & 0x7f) << shift;
    shift += 7n;
    if ((b & 0x80) === 0) break;
  }
  // Mask to 64 bits, then sign-extend.
  count = count & UINT64_MASK;
  if (count >= BIG_2_63) count -= BIG_2_64;
  return count;
}

// ---------------------------------------------------------------------------
// Decimal encoding (96-bit unscaled + sign/scale chunk)
// ---------------------------------------------------------------------------

const HALF_EVEN = 6; // decimal.js: ROUND_HALF_EVEN

/**
 * Encode a Decimal (decimal.js instance, number, or string) using the
 * T4 binary format: 1 header byte (2 bits/chunk) + up to 4 7-bit-encoded
 * magnitudes.
 * @param {Decimal | number | string} value
 * @returns {Uint8Array}
 */
export function encodeDecimal(value) {
  const d = value instanceof Decimal ? value : new Decimal(value);

  // Decompose into (sign, unscaled, scale) so that
  //   value = sign * unscaled * 10^(-scale).
  // Use toFixed() (no arg) instead of toString() so very large/small values
  // are rendered in plain decimal form rather than scientific notation —
  // BigInt() cannot parse the latter.
  const negative = d.isNegative();
  const abs = d.abs();
  const str = abs.toFixed();
  let scale = 0;
  let unscaledStr;
  const dotIdx = str.indexOf('.');
  if (dotIdx >= 0) {
    scale = str.length - dotIdx - 1;
    unscaledStr = str.slice(0, dotIdx) + str.slice(dotIdx + 1);
  } else {
    unscaledStr = str;
  }
  // Strip leading zeros (but keep at least one digit).
  unscaledStr = unscaledStr.replace(/^0+(?=\d)/, '');

  let unscaled = BigInt(unscaledStr || '0');

  // Split into three 32-bit chunks plus the sign/scale chunk.
  const lo = unscaled & 0xffff_ffffn;
  const mid = (unscaled >> 32n) & 0xffff_ffffn;
  const hi = (unscaled >> 64n) & 0xffff_ffffn;

  // priceBits[3] = scale << 16 | (0x8000_0000 if negative)
  let bits3 = BigInt(scale) << 16n;
  if (negative) bits3 |= 0x8000_0000n;

  // Treat each chunk as signed 32 for header classification.
  const bits = [lo, mid, hi, bits3].map((v) => {
    const u = Number(v); // safe: each chunk < 2^32
    return u >= 0x8000_0000 ? u - 0x1_0000_0000 : u;
  });

  // Header: 2 bits per chunk, MSB->LSB = chunk 0..3.
  let hdr = 0;
  const shifts = [6, 4, 2, 0];
  for (let i = 0; i < 4; i++) {
    const sv = bits[i];
    let tag;
    if (sv === INT32_MIN_VALUE) tag = 0x03;
    else if (sv < 0) tag = 0x02;
    else if (sv > 0) tag = 0x01;
    else tag = 0x00;
    hdr |= tag << shifts[i];
  }

  const parts = [Uint8Array.of(hdr)];
  for (let i = 0; i < 4; i++) {
    const sv = bits[i];
    if (sv !== 0 && sv !== INT32_MIN_VALUE) {
      parts.push(encode7BitInt(Math.abs(sv)));
    }
  }
  return _concat(parts);
}

/**
 * Decode a T4-binary-encoded Decimal from a stream.
 * @param {{readByte(): number}} stream
 * @returns {Decimal}
 */
export function decodeDecimal(stream) {
  const hdr = stream.readByte();
  const bits = [0, 0, 0, 0]; // signed 32

  const decodeChunk = (tag2) => {
    if (tag2 === 0x03) return INT32_MIN_VALUE;
    if (tag2 === 0x02) return -decode7BitInt(stream);
    if (tag2 === 0x01) return decode7BitInt(stream);
    return 0;
  };

  bits[0] = decodeChunk((hdr & 0xc0) >> 6);
  bits[1] = decodeChunk((hdr & 0x30) >> 4);
  bits[2] = decodeChunk((hdr & 0x0c) >> 2);
  bits[3] = decodeChunk(hdr & 0x03);

  // Reconstruct 96-bit unsigned magnitude.
  const u32 = (n) => BigInt(n >>> 0);
  const unscaled = (u32(bits[2]) << 64n) | (u32(bits[1]) << 32n) | u32(bits[0]);

  // Scale lives in bits 16..23 of chunk 3.
  const scale = (bits[3] & 0x00ff_0000) >>> 16;

  // Build the Decimal: unscaled / 10^scale, negated if chunk-3 sign bit set.
  let result;
  if (scale === 0) {
    result = new Decimal(unscaled.toString());
  } else {
    result = new Decimal(unscaled.toString()).div(new Decimal(10).pow(scale));
  }

  // Signed interpretation: bit 31 of chunk 3 => negative.
  if (bits[3] < 0) result = result.neg();

  return result;
}

// ---------------------------------------------------------------------------
// Price helpers (forward-declared, defined fully in Price.js to avoid
// a circular import here we keep them as factories accepting a Price ctor).
// ---------------------------------------------------------------------------

/**
 * Decode a Price (wraps decodeDecimal). Imported lazily to avoid a
 * cycle with price.js.
 * @param {{readByte(): number}} stream
 */
export function decodePrice(stream) {
  // Lazy import to break the cycle.
  return _lazyPrice().fromDecimal(decodeDecimal(stream));
}

/**
 * Decode a nullable Price: 1 header byte; if bit 0 is clear returns null,
 * else decodes a decimal quantized to scale 18 (half-even).
 * @param {{readByte(): number}} stream
 * @returns {object | null}
 */
export function decodePriceN(stream) {
  const hdr = stream.readByte();
  if ((hdr & 0x01) === 0x01) {
    return _lazyPrice().fromDecimal(decodeDecimal(stream));
  }
  return null;
}

// ---------------------------------------------------------------------------
// Internals
// ---------------------------------------------------------------------------

let _priceModule = null;
function _lazyPrice() {
  if (_priceModule == null) {
    // Synchronous-only callers; defer import via dynamic require pattern.
    // Top-level await isn't usable here, so we set it from Price.js itself
    // via `registerPriceFactory`.
    if (_priceFactory == null) {
      throw new Error('Price factory not registered yet (import Price.js first)');
    }
    _priceModule = { fromDecimal: _priceFactory };
  }
  return _priceModule;
}

let _priceFactory = null;
/**
 * Internal: registered by Price.js on module load to avoid a cycle.
 * @param {(d: Decimal) => object} factory
 */
export function registerPriceFactory(factory) {
  _priceFactory = factory;
  _priceModule = null;
}

function _concat(arrays) {
  let total = 0;
  for (const a of arrays) total += a.length;
  const out = new Uint8Array(total);
  let off = 0;
  for (const a of arrays) {
    out.set(a, off);
    off += a.length;
  }
  return out;
}

File: tools/Python/t4-Python-api/src/t4login/util/encoding.py
"""Port of ``com.t4login.util.EncodingUtil``.

Provides variable-length 7-bit encoding/decoding for integers, longs, and
decimals, mirroring the Java implementation byte-for-byte.

Encoding scheme (7-bit variable-length integer):
- Each byte uses 7 bits for data and the MSB as a continuation flag.
- If the MSB is set (0x80), more bytes follow; if clear, it's the last byte.
- Positive values encode in 1-5 bytes (int) or 1-9 bytes (long).
- Negative values always use the maximum size (5 bytes for int, 10 for long)
  to preserve two's complement sign semantics from Java's arithmetic right-shift.

Decimal encoding:
- A 96-bit unscaled integer is split into three 32-bit chunks (low, mid, high)
  plus a 4th chunk encoding the scale and sign.
- A 1-byte header uses 2 bits per chunk to indicate zero/positive/negative/MIN.
- Non-zero chunks are then 7-bit-encoded with their absolute values.
"""

from __future__ import annotations

import struct
from decimal import ROUND_HALF_EVEN, Decimal
from io import BytesIO
from typing import BinaryIO, TYPE_CHECKING

if TYPE_CHECKING:
    from t4login.definitions.priceconversion.price import Price

# Java int/long boundaries
_INT32_MIN = -(1 << 31)
_INT32_MAX = (1 << 31) - 1
_INT64_MIN = -(1 << 63)
_INT64_MAX = (1 << 63) - 1
_UINT32_MASK = 0xFFFF_FFFF
_UINT64_MASK = 0xFFFF_FFFF_FFFF_FFFF


# ---------------------------------------------------------------------------
# 7-bit integer encoding (mirrors Java's signed 32-bit behavior)
# ---------------------------------------------------------------------------


def encode_7bit_int(value: int) -> bytes:
    """Encode a signed 32-bit integer using variable-length 7-bit encoding.

    Positive values use 1–5 bytes (variable). Negative values always use 5 bytes
    (matching Java's arithmetic-shift behavior).
    """
    # Mask to 32-bit two's complement
    value = value & _UINT32_MASK

    if value <= _INT32_MAX:
        # Positive path: variable length
        buf = bytearray()
        while value >= 0x80:
            buf.append((value & 0xFF) | 0x80)
            value >>= 7
        buf.append(value & 0xFF)
        return bytes(buf)
    else:
        # Negative path (bit 31 set): fixed 5 bytes using arithmetic shift
        # Simulate Java's signed >> on 32-bit int
        sval = value if value < (1 << 31) else value - (1 << 32)
        buf = bytearray(5)
        buf[0] = (sval & 0xFF) | 0x80
        sval >>= 7
        buf[1] = (sval & 0xFF) | 0x80
        sval >>= 7
        buf[2] = (sval & 0xFF) | 0x80
        sval >>= 7
        buf[3] = (sval & 0xFF) | 0x80
        sval >>= 7
        buf[4] = sval & 0x0F
        return bytes(buf)


def encode_7bit_int_to_stream(value: int, out: BinaryIO) -> None:
    """Encode a signed 32-bit integer directly to a stream."""
    out.write(encode_7bit_int(value))


def decode_7bit_int(stream: BinaryIO) -> int:
    """Decode a 7-bit encoded signed 32-bit integer from a stream."""
    count = 0
    shift = 0

    while True:
        b = stream.read(1)
        if not b:
            raise EOFError("Unexpected end of stream in decode_7bit_int")
        byte_val = b[0]
        count |= (byte_val & 0x7F) << shift
        shift += 7
        if (byte_val & 0x80) == 0:
            break

    # Sign-extend from 32-bit two's complement
    if count >= (1 << 31):
        count -= 1 << 32
    return count


def decode_7bit_int_from_bytes(data: bytes) -> int:
    """Decode a 7-bit encoded integer from a byte sequence."""
    return decode_7bit_int(BytesIO(data))


# ---------------------------------------------------------------------------
# 7-bit long encoding (mirrors Java's signed 64-bit behavior)
# ---------------------------------------------------------------------------


def encode_7bit_long(value: int) -> bytes:
    """Encode a signed 64-bit integer using variable-length 7-bit encoding.

    Positive values use 1–9 bytes (variable). Negative values always use 10 bytes.
    """
    # Mask to 64-bit two's complement
    value = value & _UINT64_MASK

    if value <= _INT64_MAX:
        # Positive path: variable length
        buf = bytearray()
        while value >= 0x80:
            buf.append((value & 0xFF) | 0x80)
            value >>= 7
        buf.append(value & 0xFF)
        return bytes(buf)
    else:
        # Negative path (bit 63 set): fixed 10 bytes
        sval = value if value < (1 << 63) else value - (1 << 64)
        buf = bytearray(10)
        for i in range(9):
            buf[i] = (sval & 0xFF) | 0x80
            sval >>= 7
        buf[9] = sval & 0x0F
        return bytes(buf)


def encode_7bit_long_to_stream(value: int, out: BinaryIO) -> None:
    """Encode a signed 64-bit integer directly to a stream."""
    out.write(encode_7bit_long(value))


def decode_7bit_long(stream: BinaryIO) -> int:
    """Decode a 7-bit encoded signed 64-bit integer from a stream."""
    count = 0
    shift = 0

    while True:
        b = stream.read(1)
        if not b:
            raise EOFError("Unexpected end of stream in decode_7bit_long")
        byte_val = b[0]
        count |= (byte_val & 0x7F) << shift
        shift += 7
        if (byte_val & 0x80) == 0:
            break

    # Mask to 64 bits (Python ints are arbitrary precision; Java truncates on shift)
    count = count & _UINT64_MASK
    # Sign-extend from 64-bit two's complement
    if count >= (1 << 63):
        count -= 1 << 64
    return count


def decode_7bit_long_from_bytes(data: bytes) -> int:
    """Decode a 7-bit encoded long from a byte sequence."""
    return decode_7bit_long(BytesIO(data))


# ---------------------------------------------------------------------------
# Decimal encoding (mirrors Java's BigDecimal 96-bit split)
# ---------------------------------------------------------------------------

_INT32_MIN_VALUE = -(1 << 31)  # -2147483648


def encode_decimal(value: Decimal) -> bytes:
    """Encode a Decimal using the T4 binary format (header + 7-bit encoded chunks)."""
    out = BytesIO()
    encode_decimal_to_stream(value, out)
    return out.getvalue()


def encode_decimal_to_stream(value: Decimal, out: BinaryIO) -> None:
    """Encode a Decimal directly to a stream."""
    sign, digits, exponent = value.as_tuple()
    scale = -exponent if exponent < 0 else 0

    # Reconstruct unscaled value as positive integer (apply positive exponent as a left shift)
    unscaled = int("".join(str(d) for d in digits)) if digits else 0
    if exponent > 0:
        unscaled *= 10**exponent

    # Split into three 32-bit chunks (little-endian order: low, mid, high)
    price_bits = [0, 0, 0, 0]
    price_bits[0] = unscaled & _UINT32_MASK
    price_bits[1] = (unscaled >> 32) & _UINT32_MASK
    price_bits[2] = (unscaled >> 64) & _UINT32_MASK

    # priceBits[3] = scale << 16 | sign flag
    price_bits[3] = scale << 16
    if sign == 1:  # negative
        price_bits[3] = price_bits[3] | 0x8000_0000

    # Interpret as signed int32 for header classification
    def _as_signed32(v: int) -> int:
        v = v & _UINT32_MASK
        return v if v < (1 << 31) else v - (1 << 32)

    signed_bits = [_as_signed32(b) for b in price_bits]

    # Build header byte (2 bits per chunk)
    hdr = 0
    for i, (mask_hi, mask_lo) in enumerate([(0xC0, 6), (0x30, 4), (0x0C, 2), (0x03, 0)]):
        sv = signed_bits[i]
        if sv == _INT32_MIN_VALUE:
            bits = 0x03  # MIN_VALUE marker
        elif sv < 0:
            bits = 0x02  # negative
        elif sv > 0:
            bits = 0x01  # positive
        else:
            bits = 0x00  # zero
        hdr |= bits << mask_lo

    out.write(struct.pack("B", hdr))

    # Write non-zero, non-MIN_VALUE chunks as 7-bit encoded absolute values
    for i in range(4):
        sv = signed_bits[i]
        if sv != 0 and sv != _INT32_MIN_VALUE:
            encode_7bit_int_to_stream(abs(sv), out)


def decode_decimal(stream: BinaryIO) -> Decimal:
    """Decode a T4 binary-encoded Decimal from a stream."""
    hdr_bytes = stream.read(1)
    if not hdr_bytes:
        raise EOFError("Unexpected end of stream in decode_decimal")
    hdr = hdr_bytes[0]

    price_bits = [0, 0, 0, 0]

    # Chunk 0: bits 7-6
    tag0 = hdr & 0xC0
    if tag0 == 0xC0:
        price_bits[0] = _INT32_MIN_VALUE
    elif tag0 == 0x80:
        price_bits[0] = -1 * decode_7bit_int(stream)
    elif tag0 == 0x40:
        price_bits[0] = decode_7bit_int(stream)

    # Chunk 1: bits 5-4
    tag1 = hdr & 0x30
    if tag1 == 0x30:
        price_bits[1] = _INT32_MIN_VALUE
    elif tag1 == 0x20:
        price_bits[1] = -1 * decode_7bit_int(stream)
    elif tag1 == 0x10:
        price_bits[1] = decode_7bit_int(stream)

    # Chunk 2: bits 3-2
    tag2 = hdr & 0x0C
    if tag2 == 0x0C:
        price_bits[2] = _INT32_MIN_VALUE
    elif tag2 == 0x08:
        price_bits[2] = -1 * decode_7bit_int(stream)
    elif tag2 == 0x04:
        price_bits[2] = decode_7bit_int(stream)

    # Chunk 3: bits 1-0
    tag3 = hdr & 0x03
    if tag3 == 0x03:
        price_bits[3] = _INT32_MIN_VALUE
    elif tag3 == 0x02:
        price_bits[3] = -1 * decode_7bit_int(stream)
    elif tag3 == 0x01:
        price_bits[3] = decode_7bit_int(stream)

    # Reconstruct 96-bit unscaled integer from three unsigned 32-bit chunks
    int2 = (price_bits[2] & _UINT32_MASK) << 32
    int1 = (int2 + (price_bits[1] & _UINT32_MASK)) << 32
    integer = int1 + (price_bits[0] & _UINT32_MASK)

    # Extract scale from chunk 3 (bits 16-23)
    scale = (price_bits[3] & 0xFF0000) >> 16

    # Build decimal
    result = Decimal(integer) / Decimal(10 ** scale) if scale > 0 else Decimal(integer)

    # Negate if bit 31 of priceBits[3] is set
    if price_bits[3] < 0:  # signed interpretation — bit 31 set means negative
        result = -result

    return result


def decode_decimal_from_bytes(data: bytes) -> Decimal:
    """Decode a T4 binary-encoded Decimal from a byte sequence."""
    return decode_decimal(BytesIO(data))


# ---------------------------------------------------------------------------
# Price decoding helpers
# ---------------------------------------------------------------------------


def decode_price(stream: BinaryIO) -> "Price":
    """Decode a Price from a stream (wraps decode_decimal in a Price object)."""
    from t4login.definitions.priceconversion.price import Price

    return Price(decode_decimal(stream))


def decode_price_n(stream: BinaryIO) -> "Price | None":
    """Decode a nullable Price from a stream.

    Reads 1 header byte; returns None if bit 0 is clear, otherwise decodes
    the decimal and quantizes to scale 18 with HALF_EVEN rounding.
    """
    from t4login.definitions.priceconversion.price import Price, Scale

    hdr_bytes = stream.read(1)
    if not hdr_bytes:
        raise EOFError("Unexpected end of stream in decode_price_n")
    hdr = hdr_bytes[0]

    if (hdr & 0x01) == 0x01:
        dec = decode_decimal(stream)
        quantum = Decimal(10) ** -Scale
        dec = dec.quantize(quantum, rounding=ROUND_HALF_EVEN)
        return Price(dec)
    else:
        return None
